src/main/java/com/programmers/springbootbasic/domain/voucher/Voucher.java

```diff
@@ -0,0 +1,85 @@
+package com.programmers.springbootbasic.domain.voucher;
+
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+public abstract class Voucher {
+    private final UUID voucherId;
+    private final String name;
+    private final Long minimumPriceCondition;
+    private final LocalDateTime expirationDate;
+    private VoucherState state = VoucherState.AVAILABLE;
+
+    protected Voucher(UUID voucherId, String name, LocalDateTime expirationDate) {
+        if (isInvalidVoucherId(voucherId)) {
+            throw new IllegalArgumentException("잘못된 아이디 값");
+        }
+        this.voucherId = voucherId;
+        if (isInvalidName(name)) {
+            throw new IllegalArgumentException("잘못된 이름");
+        }
+        this.name = name;
+        this.minimumPriceCondition = 0L;
+        if (isInvalidExpirationDate(expirationDate)) {
+            throw new IllegalArgumentException("잘못된 유효기간");
+        }
+        this.expirationDate = expirationDate;
+    }
+
+    protected Voucher(UUID voucherId, String name, Long minimumPriceCondition, LocalDateTime expirationDate) {
+        if (isInvalidVoucherId(voucherId)) {
+            throw new IllegalArgumentException("잘못된 아이디 값");
+        }
+        this.voucherId = voucherId;
+        if (isInvalidName(name)) {
+            throw new IllegalArgumentException("잘못된 이름");
+        }
+        this.name = name;
+        this.minimumPriceCondition = minimumPriceCondition;
+        if (isInvalidExpirationDate(expirationDate)) {
+            throw new IllegalArgumentException("잘못된 유효기간");
+        }
+        this.expirationDate = expirationDate;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public Long getMinimumPriceCondition() {
+        return minimumPriceCondition;
+    }
+
+    public LocalDateTime getExpirationDate() {
+        return expirationDate;
+    }
+
+    public VoucherState getState() {
+        return state;
+    }
+
+    private boolean isInvalidVoucherId(UUID voucherId) {
+        return voucherId == null;
+    }
+
+    private boolean isInvalidName(String name) {
+        return name == null || name.isEmpty();
+    }
+
+    private boolean isInvalidExpirationDate(LocalDateTime expirationDate) {
+        LocalDateTime now = LocalDateTime.now();
+        return now.isAfter(expirationDate);
+    }
+
+    protected boolean isExpiration(LocalDateTime now) {
+        return now.isAfter(expirationDate);
+    }
+
+    protected void setVoucherState(VoucherState toChange) {
+        state = toChange;
+    }
+
+    protected abstract Long getDiscountPrice(Long priceBeforeDiscount);
+
+    protected abstract Long discount(Long priceBeforeDiscount);
```

Comment:
`discount`는 할인 금액을 반환하고 바우처의 상태를 변경하는 것 같은데 추상 메서드가 아니라  부모 클래스에서 기본 구현을 제공하는 방식이 나을 것 같다는 생각이 드네요

```suggestion
    public Long discount(Long priceBeforeDiscount) {
        this.state = VoucherState.USED;
        return getDiscountPrice(priceBeforeDiscount);
    }

    protected abstract Long getDiscountPrice(Long priceBeforeDiscount);
```

Reply:
discount 부모클래스에서 구현했습니다.

[d79ba90](https://github.com/prgrms-be-devcourse/springboot-basic/pull/660/commits/d79ba904c07e7a10a6ebd56d96e60980f1f4ad28)